packages/run-protocol/src/runStake/attestation.js

```diff
@@ -0,0 +1,256 @@
+
+import { AmountMath, AssetKind } from '@agoric/ertp';
+import { E, Far } from '@endo/far';
+import { fit, M, makeCopyBag, makeStore } from '@agoric/store';
+import { assertProposalShape } from '@agoric/zoe/src/contractSupport/index.js';
+
+const { details: X } = assert;
+
+export const KW = harden({
+  /** seat keyword for use in offers to return an attestation. */
+  Attestation: 'Attestation',
+});
+
+/**
+ * Find-or-create value in store.
+ *
+ * @type {<K, V>(store: Store<K, V>, key: K, thunk: () => V) => V}
+ */
+const getOrElse = (store, key, make) => {
+  if (store.has(key)) {
+    return store.get(key);
+  }
+  const value = make();
+  store.init(key, value);
+  return value;
+};
+
+/**
+ * To support `makeAttestation`, we directly mint attestation payments.
+ * We still use a ZCFMint because returning attestations is done
+ * through Zoe offers.
+ *
+ * @param {ZCF} zcf
+ * @param {ZCFMint} zcfMint
+ * @param {Amount} amountToMint
+ * @returns {Promise<Payment>}
+ */
+const mintZCFMintPayment = (zcf, zcfMint, amountToMint) => {
+  const { userSeat, zcfSeat } = zcf.makeEmptySeatKit();
+  zcfMint.mintGains(harden({ [KW.Attestation]: amountToMint }), zcfSeat);
+  zcfSeat.exit();
+  return E(userSeat).getPayout(KW.Attestation);
+};
+
+/**
+ * Make an ZCFMint attentuated so that minting represents
+ * putting a lien on staked assets. Payments can be returned,
+ * releasing (part of) the lien.
+ *
+ * @param {ZCF} zcf
+ * @param {string} keyword for use in makeZCFMint
+ * @param {Brand<'nat'>} stakeBrand brand of the staked assets
+ * @param {ERef<StakingAuthority>} lienBridge bridge to account state
+ */
+const makeAttestationKit = async (zcf, keyword, stakeBrand, lienBridge) => {
+  const { add, subtract, makeEmpty, isGTE, coerce } = AmountMath;
+  const empty = makeEmpty(stakeBrand);
+  const zcfMint = await zcf.makeZCFMint(keyword, AssetKind.COPY_BAG);
+  const { issuer, brand: attBrand } = zcfMint.getIssuerRecord();
+
+  /**
+   * @param {string} address
+   * @param {Amount<'nat'>} lienedAmount
+   * @returns {Amount<'copyBag'>}
+   */
+  const wrapLienedAmount = (address, lienedAmount) =>
+    AmountMath.make(attBrand, makeCopyBag([[address, lienedAmount.value]]));
+
+  /** @param {Amount<'copyBag'>} attestationAmount */
+  const unwrapLienedAmount = attestationAmount => {
+    assert(
+      attestationAmount.brand === attBrand,
+      X`The escrowed attestation ${attestationAmount} was not of the attestation brand ${attBrand}`,
+    );
+    // Caller is responsible to split payload of >1 item.
+    fit(attestationAmount.value.payload, harden([[M.string(), M.bigint()]]));
+    /** @type {[string, bigint][]} */
+    const [[address, valueReturned]] = attestationAmount.value.payload;
+    const lienedAmount = AmountMath.make(stakeBrand, valueReturned);
+    return { address, lienedAmount };
+  };
+
+  /** @type {(hi: Amount, lo: Amount) => Amount} */
+  const subtractOr0 = (hi, lo) => (isGTE(hi, lo) ? subtract(hi, lo) : empty);
```

Comment:
hax0r codez!

jk/ but consider `subtractOrEmpty` or `floorSubtract`.

also, param names are confusing since `lo` may be greater than `hi`.  I had to look it up, but apparently there are mathematical terms: Minuend - Subtrahend = Difference.

```suggestion
  const subtractToZero = (m, s) => (isGTE(m, s) ? subtract(m, s) : empty);
```

Reply:
The odds that someone is going to recognize `m` and `s` as short for `minuend` and `subtrahend` seem remote. I was going to use `a` and `b` or `x` and `y`, but I thought `hi` and `lo` were a little more clear: even though `lo` could be higher than `hi`, that's the exceptional case that rounds to `empty`.

The term I found was _truncated subtraction_ ([source](https://math.stackexchange.com/a/328542)). But that introduces a line-wrap, which (to quote Dean) is evil. But `subtractOrEmpty` works with 1-letter args. OK. Done.